legend-depot-core-tracing/src/main/java/org/finos/legend/depot/tracing/resources/BaseResource.java

```diff
@@ -97,8 +101,67 @@ protected <T> T handle(String resourceAPIMetricName, String label, Supplier<T> s
         }
     }
 
+    protected <T> Response handle(String resourceAPIMetricName, String label, Supplier<T> supplier, Request request, String groupId, String artifactId, String versionId)
+    {
+        return handle(resourceAPIMetricName, label, supplier, request, groupId, artifactId, versionId, null, null);
+    }
+
+    protected <T> Response handle(String resourceAPIMetricName, String label, Supplier<T> supplier, Request request, String groupId, String artifactId, String versionId, String clientVersion, String serverVersion)
+    {
+        EntityTag serverTag = generateETag(groupId, artifactId, versionId, clientVersion, serverVersion);
+        if (serverTag != null && request != null && request.evaluatePreconditions(serverTag) != null)
+        {
+            return Response.noContent().status(Response.Status.NOT_MODIFIED).build();
+        }
+        try
+        {
+            T output = handle(resourceAPIMetricName, label, supplier);
+            return Response.ok(output).tag(serverTag).build();
+        }
+        catch (Exception e)
+        {
+            throw new WebApplicationException(e.getMessage(), e);
+        }
+    }
+
     protected <T> T handle(String label, Supplier<T> supplier)
     {
         return handle(label, label, supplier);
     }
+
+    protected <T> Response handle(String label, Supplier<T> supplier, Request request, String groupId, String artifactId, String versionId)
+    {
+        return handle(label, label, supplier, request, groupId, artifactId, versionId);
+    }
+
+    protected <T> Response handle(String label, Supplier<T> supplier, Request request, String groupId, String artifactId, String versionId, String clientVersion, String serverVersion)
+    {
+        return handle(label, label, supplier, request, groupId, artifactId, versionId, clientVersion, serverVersion);
+    }
+
+    protected EntityTag generateETag(String groupId, String artifactId, String versionId, String clientVersion, String serverVersion)
+    {
+        if (!versionId.endsWith("-SNAPSHOT"))
+        {
+            StringBuilder etagBuilder = new StringBuilder().append(Objects.hashCode(groupId + ":" + artifactId + ":" + versionId));
```

Comment:
etag calculation is different per api so you cant mingle all params together or all logic here.
 the best way to achieve this is to pass Supplier function( ie Supplier<String> etagSupplier) and let the client pass the function so you only do eval here.
that would also clean up the api as params are variable